### PR TITLE
Add SupportUser class and use implicit association class casting

### PR DESCRIPTION
**Pull request by Nitemaeric**
## Context

As [STI has been adopted for the User class](https://github.com/DFE-Digital/itt-mentor-services/pull/168), we can now fully lean into the practice of using only namespaced models and with Rails' _magic_ ✨, implicit namespaces for associations.

## Changes proposed in this pull request

- Drop `.claims` and `.placements` scopes from User. A User can be one or the other, never both, therefore we can rely on calling `Claims::User` and `Placements::User` for `User.claims` and `User.placements`.

## Guidance to review

  ```ruby
  class Claims::User < User
    has_many :schools # This will use the current namespace and look for Claims::School by default.
  end
  ```

- Try going through flows. Nothing should break.
- Try the following in your console:

  ```ruby
  # To silence SQL query logs
  ApplicationRecord.logger.silence do
    [
      Claims::User.all.flat_map(&:schools).all? { _1.class == Claims::School },
      Claims::School.all.flat_map(&:users).all? { _1.class == Claims::User },
      Placements::User.all.flat_map(&:schools).all? { _1.class == Placements::School },
      Placements::User.all.flat_map(&:providers).all? { _1.class == Placements::Provider },
      Placements::School.all.flat_map(&:users).all? { _1.class == Placements::User },
      Placements::Provider.all.flat_map(&:users).all? { _1.class == Placements::User },
    ]
  end
  ```

-----

Whilst trying to fix the `config/initializers/preload_stis.rb` code running without a database, I stumbled on optimising our GitHub workflow. Lint is around 50% faster without setting up Node and more.

| Before | After |
| ------ | ----- |
| CleanShot 2024-01-29 at 15 30 05 | CleanShot 2024-01-29 at 15 31 36 |

### CODE REVIEW

File: config/initializers/preload_stis.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#option-4-preload-types-from-the-database
+
+if Rails.env.development? && Rails.application.config.eager_load
+  Rails.application.config.to_prepare do
+    ApplicationRecord.logger.silence do
+      types = User.unscoped.select(:type).distinct.pluck(:type)
+      types.compact.each(&:constantize)
+    end
+  end
+end
```
Comment (ollietreend) [inline config/initializers/preload_stis.rb:10]: I'm not sure why this is needed. I've read the linked docs, but I'm still not sure what the need is for preloading the different user classes.

When I check out the branch locally and run this in the Rails console:

```ruby
User.first.class.name
  User Load (0.3ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> "Claims::SupportUser"
```

I get the expected result, and no weird log output, regardless of whether I have this initializer or not. It seems to work the same either way? 🤔 

Can you ELI5? 🙏🏻

File: app/models/user.rb [hunk 2]
```diff
@@ -33,4 +33,8 @@ def full_name
   def support_user?
     false
   end
+
+  def service
+    self.class.name.split("::").first.downcase.to_sym
+  end
```
Comment (ollietreend) [inline app/models/user.rb:36]: A comment could be nice here, because this line seems quite difficult to parse:

```suggestion
  def service
    # Infer service from class namespace – e.g. Placements::User becomes :placements
    self.class.name.split("::").first.downcase.to_sym
  end
```

File: app/models/claims/school.rb [hunk 3]
```diff
@@ -47,6 +47,6 @@
 class Claims::School < School
   default_scope { claims_service }
 
-  has_many :users, -> { claims }, through: :memberships
+  has_many :users, through: :memberships
```
Comment (ollietreend) [inline app/models/claims/school.rb:50]: Nice!

File: spec/models/claims/school_spec.rb [hunk 4]
```diff
@@ -49,6 +49,21 @@
...
+        claims_school = create(:claims_school)
+        claims_user = create(:claims_user)
+
+        claims_school.users << claims_user
+        claims_school.users << create(:placements_user)
+
+        expect(claims_school.users).to contain_exactly(claims_user)
+        expect(claims_school.users).to all(be_a(Claims::User))
+      end
+    end
```
Comment (ollietreend) [inline spec/models/claims/school_spec.rb:66]: Nice test coverage, too.

**Review by ollietreend — COMMENTED** [review]
Really nice stuff 🎉

I think we've really hit upon something good with #164, #168 and now this PR.

Just one question from me, and one comment.

File: config/initializers/preload_stis.rb (hunk 1 above)
Comment (Nitemaeric) [reply config/initializers/preload_stis.rb:10]: I messed up inverting the latter condition... It should be `if Rails.env.development? && !Rails.application.config.eager_load`

Anyway...

Try running `SupportUser.all` without this code. It will default to `where(type: "SupportUser")` in development as it isn't aware of the additional subclasses that inherit from `SupportUser`.

File: app/models/user.rb (hunk 2 above)
Comment (Nitemaeric) [reply app/models/user.rb]: Updated. Added examples and method documentation using standard RDoc/YARD documentation format.

**Review by ollietreend — APPROVED** [review]
Nice work ✨ 🚀
